### PR TITLE
[core] Drain actor's references before exiting

### LINKED ISSUE #17718: [core] Actors force-killed while they still own objects that are referenced
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):* 2.0dev

Actors are ref-counted and will get force-killed if there are no more references to the actor.

However, this can be problematic if another process references an object owned by that actor. Then, that reference becomes unusable because the actor has been killed. We should fix this by only exiting the actor once its refs have drained.

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments): See #17717.

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

When actors go out of scope, they are currently force-killed by the GCS. Unfortunately this means that if they own any objects that are still in scope by another process, the other process will not be able to use their ObjectRef anymore (#17718).

This fixes it by, (a) gentle-killing actors when they go out of scope, and (b) for gentle kills only, actor's owned references are drained before exiting the process.

This also changes the reference draining condition to be weaker than before: previously we wouldn't exit the worker if any object was still owned. This can lead to worker leaks, e.g., an actor will stay alive forever if it calls something like `self.lifetime_ref = ray.put(...)`. Now the worker will only stay alive if there are any objects that are owned *and in reference by a different worker*.

## Related issue number

Closes #17718.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 1]
```diff
@@ -56,7 +56,6 @@ class CoreWorkerDirectActorTaskSubmitterInterface {
   virtual void DisconnectActor(
       const ActorID &actor_id, int64_t num_restarts, bool dead,
       const std::shared_ptr<rpc::RayException> &creation_task_exception = nullptr) = 0;
-  virtual void KillActor(const ActorID &actor_id, bool force_kill, bool no_restart) = 0;
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_actor_transport.h:59]: Also removed some unused code for killing actors.

**Review by ericl — COMMENTED** [review]
Don't we still need the force kill logic for manual actor force kills?

**Comment by ericl** [comment]
Seems tests are failing.

**Comment by stephanie-wang** [comment]
> Don't we still need the force kill logic for manual actor force kills?

It's not used (the GCS kills actors now).

**Comment by stephanie-wang** [comment]
Closing for now, but we should revisit this to close #17718. Most of this code can be reused, but we should focus on usability and better warnings instead of trying to wait to drain an actor's references.

**Pull request closed without merging**